pkg/apis/core/types.go

```diff
@@ -1966,6 +1966,33 @@ const (
 	PullIfNotPresent PullPolicy = "IfNotPresent"
 )
 
+// ContainerResizePolicy specifies user guidance on how container resource resize should be handled.
+// Only one of the following container resize policies may be specified.
+// If none of the following policies is specified, it defaults to NoRestart.
+type ContainerResizePolicy string
+
+// These are the valid container resize policies:
+// NoRestart policy tells Kubelet to call UpdateContainerResources CRI API to resize
+// the resources without restarting the container, if possible. This is the default behavior.
+// RestartContainer policy tells Kubelet to stop and start the container with when
+// new resources are applied. This is needed for legacy applications e.g. java apps
+// using -xmxN flag which are unable to use the resized memory without restarting.
+const (
+	// Resize the container in-place without restarting it.
+	NoRestart ContainerResizePolicy = "NoRestart"
+	// Resize the container in-place by restarting it after resize.
+	RestartContainer ContainerResizePolicy = "RestartContainer"
+)
+
+// ResizePolicy represents the resource resize policy for a single container.
+type ResizePolicy struct {
```

Comment:
ResourceList is one of the only places we use `map` in the API, but it fits well.  There's a value to consistency - why is this better as a struct than a map?

Reply:
@thockin I wanted this to be a map ([KEP discussion](https://github.com/kubernetes/enhancements/pull/686/#discussion_r320026990)) for simplicity (both for user and for code) despite the API conventions recommending [lists of named subobjects preferred over maps](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps) .. cpu and memory are the limited key choices here vs container object where user can choose the name.

We discussed this during [API live review in 2019 KubeCon San Diego](https://www.youtube.com/watch?v=faRARV3C7Fk&feature=youtu.be&t=3467), and we leaned towards staying with the API conventions as it makes future extensibility simple. 

@liggitt Do you think we can revisit this? Can we make an exception here and use maps even though at this point we don't have any reason for the values to take parameters? I haven't really spent a lot of time seeing APIs evolve, but if you feel there are strong reasons for value to be objects instead of string (my preference because it is simple), we can do that. Are we looking at something like [ContainerState](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L2179)?

Reply:
If we make the "value" a struct, then a list probably makes more sense.

If we make the "value" a single value, I have a slight lean towards the map, for congruence with the `resources` field

Reply:
Then we should leave it as a list - it is possible future iterations may want to qualify the behavior a bit more, such as number of failed update retry attempts before restarting.